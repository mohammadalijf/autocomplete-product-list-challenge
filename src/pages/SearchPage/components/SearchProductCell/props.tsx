import { ICSVProducts } from "../../../../services/productsAPI";

export type SearchProductCellProps = {
  /**
   * product item to show in cell
   */
  product: ICSVProducts;
};
