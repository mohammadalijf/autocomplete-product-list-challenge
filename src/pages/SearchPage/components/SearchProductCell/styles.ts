import { makeStyles } from "@material-ui/core";

const useSearchProductCellStyles = makeStyles((theme) => ({
  container: {
    position: "relative",
    boxShadow: theme.shadows[1],
    borderRadius: theme.shape.borderRadius,
    background: theme.palette.background.paper,
    padding: theme.spacing(1),
    overflow: "hidden",
    display: "grid",
    gridGap: theme.spacing(1),
    justifyItems: "center",
    transition: "grid-column 1s",
  },
  titleLabel: {
    fontWeight: theme.typography.fontWeightBold,
    fontSize: theme.typography.fontSize,
    width: "100%",
  },
  inforContainer: {
    width: "100%",
  },
  infoTitleLabel: {
    fontWeight: theme.typography.fontWeightBold,
    marginRight: theme.spacing(1),
    width: "10ch",
    display: "inline-block",
    whiteSpace: "nowrap",
  },
  imgContainer: {
    display: "flex",

    justifyContent: "center",
    flexWrap: "wrap",
  },
  img: {
    width: "20ch",
    height: "28.87ch",
    objectFit: "contain",
    margin: theme.spacing(1),
  },
  saleBadge: {
    backgroundColor: theme.palette.error.main,
    color: theme.palette.background.paper,
    fontWeight: theme.typography.fontWeightBold,
    fontSize: theme.typography.caption.fontSize,
    textAlign: "center",
    position: "absolute",
    padding: theme.spacing(1),
    right: -29,
    top: 11,
    width: 110,
    transform: "rotate(45deg)",
  },
  expanded: {
    gridColumn: "1 / -1",
  },
}));

export default useSearchProductCellStyles;
