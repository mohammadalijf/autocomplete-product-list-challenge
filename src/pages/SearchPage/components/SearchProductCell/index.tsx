import clsx from "clsx";
import { Button, Typography } from "@material-ui/core";
import { FunctionComponent, useEffect, useRef, useState } from "react";
import { SearchProductCellProps } from "./props";
import useSearchProductCellStyles from "./styles";

/**
 * Component to display products in a cell
 */
const SearchProductCell: FunctionComponent<SearchProductCellProps> = ({
  product,
  ...props
}) => {
  const classes = useSearchProductCellStyles();
  const containerRef = useRef<HTMLDivElement>(null);

  //#region states

  // boolean for cell expansion state
  const [expanded, setExpanded] = useState<boolean | undefined>(undefined);
  // boolean for showing image instead of placeholder
  const [loadImg, setLoadImg] = useState(false);

  //#endregion

  //#region handlers

  /**
   * handler for OnClick event of "show more image" button
   */
  const showMoreImageClickHandler = () => {
    setExpanded((expanded) => !expanded);
  };

  //#endregion

  //#region side effects

  // scroll item to center of viewport on expanded changes
  useEffect(() => {
    if (expanded === undefined) return;
    containerRef.current?.scrollIntoView({
      behavior: "smooth",
      block: "center",
    });
  }, [expanded]);

  // track if image is visible in viewport for image lazy loading
  useEffect(() => {
    const scrollHandler = () => {
      const boundingClientRect = containerRef.current?.getBoundingClientRect();
      const height = containerRef.current?.offsetHeight;
      const width = containerRef.current?.offsetWidth;
      if (!boundingClientRect || !height || !width) return;

      const viewportWidth =
        window.innerWidth || document.documentElement.clientWidth;

      const viewportHeight =
        window.innerHeight || document.documentElement.clientHeight;

      const isInViewport =
        boundingClientRect.top >= -height &&
        boundingClientRect.left >= -width &&
        boundingClientRect.right <= viewportWidth + width &&
        boundingClientRect.bottom <= viewportHeight + height;

      if (isInViewport) {
        window.removeEventListener("scroll", scrollHandler);
      }
      setLoadImg((imgLoad) => imgLoad || isInViewport);
    };
    window.addEventListener("scroll", scrollHandler);
    scrollHandler();
    return () => {
      window.removeEventListener("scroll", scrollHandler);
    };
  }, [setLoadImg]);

  //#endregion

  return (
    <article
      data-testid="productCell"
      ref={containerRef}
      className={clsx(classes.container, { [classes.expanded]: expanded })}
    >
      <div className={classes.imgContainer}>
        <img
          data-testid="productMainImg"
          loading="lazy"
          alt={product.title}
          src={loadImg ? product.imageLink : "/placeholder.png"}
          className={classes.img}
        />
        {expanded &&
          product.additionalImageLink.map((image) => (
            <img
              key={image}
              alt={product.title}
              src={image}
              className={classes.img}
              loading="lazy"
            />
          ))}
      </div>
      <Typography
        component="h6"
        className={classes.titleLabel}
        data-testid="productTitle"
      >
        {product.title}
      </Typography>
      <div className={classes.inforContainer}>
        <div>
          <Typography variant="caption" className={classes.infoTitleLabel}>
            GTIN
          </Typography>
          <Typography variant="caption" data-testid="productGTIN">
            {product.gtin}
          </Typography>
        </div>
        <div>
          <Typography variant="caption" className={classes.infoTitleLabel}>
            Gender
          </Typography>
          <Typography variant="caption" data-testid="productGender">
            {product.gender}
          </Typography>
        </div>
        <div>
          <Typography variant="caption" className={classes.infoTitleLabel}>
            Price
          </Typography>
          <Typography variant="caption" data-testid="productPrice">
            {product.price}
          </Typography>
        </div>
        <div>
          <Typography variant="caption" className={classes.infoTitleLabel}>
            Sale Price
          </Typography>
          <Typography variant="caption" data-testid="productSalePrice">
            {product.salePrice}
          </Typography>
        </div>
      </div>
      {product.additionalImageLink.length > 0 && (
        <Button
          data-testid="moreImageButton"
          variant="outlined"
          color="primary"
          onClick={showMoreImageClickHandler}
        >
          {expanded ? "show less images" : "show more images"}
        </Button>
      )}
      {product.onSale && (
        <div className={classes.saleBadge} data-testid="saleLabel">
          On Sale
        </div>
      )}
    </article>
  );
};

export default SearchProductCell;
