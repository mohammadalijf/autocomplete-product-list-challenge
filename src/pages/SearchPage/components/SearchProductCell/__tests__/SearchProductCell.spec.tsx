import { cleanup, render, RenderResult } from "@testing-library/react";
import SearchProductCell from "..";
import {
  CSVProductGender,
  ICSVProducts,
} from "../../../../../services/productsAPI";

describe("<SearchProductCell />", () => {
  const product: ICSVProducts = {
    title: "Jeans",
    gtin: "1",
    price: "20.0 EUR",
    salePrice: "20.0 EUR",
    gender: CSVProductGender.Female,
    imageLink: "",
    additionalImageLink: [],
  };
  let searchProductGrid: RenderResult;

  afterEach(cleanup);

  beforeEach(() => {
    searchProductGrid = render(<SearchProductCell product={product} />);
  });

  it("renders correctly", () => {
    expect(searchProductGrid.asFragment()).toMatchSnapshot();
  });

  it("should show correct information", () => {
    expect(searchProductGrid.getByTestId("productTitle").textContent).toEqual(
      product.title
    );

    expect(searchProductGrid.getByTestId("productGTIN").textContent).toEqual(
      product.gtin
    );

    expect(searchProductGrid.getByTestId("productPrice").textContent).toEqual(
      product.price
    );

    expect(
      searchProductGrid.getByTestId("productSalePrice").textContent
    ).toEqual(product.salePrice);

    expect(searchProductGrid.getByTestId("productGender").textContent).toEqual(
      product.gender
    );

    expect(searchProductGrid.getByRole("img").getAttribute("src")).toEqual(
      "/placeholder.png"
    );

    expect(searchProductGrid.queryAllByTestId("moreImageButton")).toHaveLength(
      0
    );
  });
});
