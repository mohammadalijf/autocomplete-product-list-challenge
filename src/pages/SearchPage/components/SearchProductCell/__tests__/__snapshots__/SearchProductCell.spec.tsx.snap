// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchProductCell /> renders correctly 1`] = `
<DocumentFragment>
  <article
    class="makeStyles-container-1"
    data-testid="productCell"
  >
    <div
      class="makeStyles-imgContainer-5"
    >
      <img
        alt="Jeans"
        class="makeStyles-img-6"
        loading="lazy"
        src="/placeholder.png"
      />
    </div>
    <h6
      class="MuiTypography-root makeStyles-titleLabel-2 MuiTypography-body1"
      data-testid="productTitle"
    >
      Jeans
    </h6>
    <div
      class="makeStyles-inforContainer-3"
    >
      <div>
        <span
          class="MuiTypography-root makeStyles-infoTitleLabel-4 MuiTypography-caption"
        >
          GTIN
        </span>
        <span
          class="MuiTypography-root MuiTypography-caption"
          data-testid="productGTIN"
        >
          1
        </span>
      </div>
      <div>
        <span
          class="MuiTypography-root makeStyles-infoTitleLabel-4 MuiTypography-caption"
        >
          Gender
        </span>
        <span
          class="MuiTypography-root MuiTypography-caption"
          data-testid="productGender"
        >
          female
        </span>
      </div>
      <div>
        <span
          class="MuiTypography-root makeStyles-infoTitleLabel-4 MuiTypography-caption"
        >
          Price
        </span>
        <span
          class="MuiTypography-root MuiTypography-caption"
          data-testid="productPrice"
        >
          20.0 EUR
        </span>
      </div>
      <div>
        <span
          class="MuiTypography-root makeStyles-infoTitleLabel-4 MuiTypography-caption"
        >
          Sale Price
        </span>
        <span
          class="MuiTypography-root MuiTypography-caption"
          data-testid="productSalePrice"
        >
          20.0 EUR
        </span>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
